observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingChannelInterceptor.java

```diff
@@ -16,33 +16,248 @@
+import com.google.protobuf.util.Durations;
 import io.grpc.CallOptions;
 import io.grpc.Channel;
 import io.grpc.ClientCall;
 import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Deadline;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
 import io.grpc.Internal;
+import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventLogger;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
-/** A logging interceptor for {@code LoggingChannelProvider}. */
+/**
+ * A logging interceptor for {@code LoggingChannelProvider}.
+ */
 @Internal
 public final class InternalLoggingChannelInterceptor implements ClientInterceptor {
+  private static final Logger logger = Logger
+      .getLogger(InternalLoggingChannelInterceptor.class.getName());
+
+  private final LogHelper helper;
 
   public interface Factory {
     ClientInterceptor create();
   }
 
   public static class FactoryImpl implements Factory {
+    private final Sink sink;
+    private final LogHelper helper;
+
+    static LogHelper createLogHelper(Sink sink, TimeProvider provider) {
+      return new LogHelper(sink, provider);
+    }
+
+    public FactoryImpl(Sink sink) {
+      this.sink = sink;
+      this.helper = createLogHelper(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
+    }
 
     @Override
     public ClientInterceptor create() {
-      return new InternalLoggingChannelInterceptor();
+      return new InternalLoggingChannelInterceptor(helper);
+    }
+
+    public void close() {
+      sink.close();
     }
   }
 
+  private InternalLoggingChannelInterceptor(LogHelper helper) {
+    this.helper = helper;
+  }
+
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
       CallOptions callOptions, Channel next) {
-    // TODO(dnvindhya) implement the interceptor
-    return null;
+
+    final AtomicLong seq = new AtomicLong(1);
+    final String rpcId = UUID.randomUUID().toString();
+    final String authority = next.authority();
+    final String serviceName = method.getServiceName();
+    final String methodName = method.getBareMethodName();
+    // Get the stricter deadline to calculate the timeout once the call starts
+    final Deadline deadline = LogHelper.min(callOptions.getDeadline(),
+        Context.current().getDeadline());
+
+    // TODO (dnvindhya): implement isMethodToBeLogged() to check for methods to be logged
+    // according to config. Until then always return true.
+    if (!helper.isMethodToBeLogged(method.getFullMethodName())) {
```

Comment:
I understand this is as per the previous suggestion. Have you thought about unit testing this code path?

Reply:
Since the methods always return `true`, it logs all methods. 
Added TODO to unit test once logging filters are implemented.

Reply:
I was more thinking about using a "spy" on the helper (since mocking the helper is not possible with the current structure) and returning `false` through the spy to test the false code path. But never mind - we will do that in a separate PR